### PR TITLE
Add new state transfer tests for validating the RVT's root validity among replicas

### PR DESCRIPTION
**Pull request by StefanPalashev**
This PR introduces three new tests that validate the Range validation trees (RVTs) of the replicas are synchronized after specific scenarios.
* test_state_transfer_rvt_validity - after running the consensus for 10 checkpoints.
* test_state_transfer_rvt_validity_after_pruning - after running the consensus and then pruning.
* test_state_transfer_rvt_root_validation_after_adding_blocks - after running the consensus multiple times while there are random restarts and prunings.

All of the previous tests in this test suite are now also validating the RVTs.

The validation is done based on a new function `wait_for_replicas_rvt_root_values_to_be_in_sync`.

### CODE REVIEW

**Review by mdarade — APPROVED** [review]
Looks Good.
Thanks Stefan.

**Comment by glevkovich** [comment]
@StefanPalashev , You need to reword one of the commits. It has a header "wp" ?
At the same time, please pass on all commits and make sure the comment and headers make sense.

File: tests/apollo/util/bft.py [hunk 1]
```diff
@@ -1256,6 +1256,46 @@ async def wait_for_state_transfer_to_stop_with_RFMD(self, up_to_date_node, stale
                                     action.add_success_fields(n=n, expected_seq_num=expected_seq_num)
                                     return
 
+    async def wait_for_replicas_rvt_root_values_to_be_in_sync(self, replica_ids):
```
Comment (glevkovich) [inline tests/apollo/util/bft.py]: nit: have the waiting time as a parameter, put 30 seconds to be default.

File: tests/apollo/util/bft.py [hunk 2]
```diff
@@ -1256,6 +1256,46 @@ async def wait_for_state_transfer_to_stop_with_RFMD(self, up_to_date_node, stale
...
+                    print(root_values)
+                    # At this point all replicas' root values are collected
+                    if root_values.count(root_values[0]) == len(root_values) and len(root_values[0]) > 0:
+                        break
+                    else:
+                        await trio.sleep(1)
+
+    async def wait_for_rvt_root_value(self, replica_id, root_values):
+        """
+        Wait for a single replica to return the current value of the root of its Range validation tree.
```
Comment (glevkovich) [inline tests/apollo/util/bft.py:1285]: nit: explain wheat is the returned value

File: tests/apollo/util/bft.py [hunk 3]
```diff
@@ -1256,6 +1256,46 @@ async def wait_for_state_transfer_to_stop_with_RFMD(self, up_to_date_node, stale
...
+        Check every .5 seconds and fail after 30 seconds.
+        """
+        with log.start_action(action_type="wait_for_rvt_root_value", replica=replica_id) as action:
```
Comment (glevkovich) [inline tests/apollo/util/bft.py]: I'm not sure why are you putting the eliot_logging here? I might be missing something. Shouldn't it be under the actual call (line 1297)? This is the definition.

File: tests/apollo/test_skvbc_state_transfer.py [hunk 4]
```diff
@@ -186,3 +192,84 @@ async def test_state_transfer_for_two_successive_cycles(self, bft_network,exchan
...
+
+        latest_pruneable_blocks = []
+        rsi_rep = client.get_rsi_replies()
+        for r in rsi_rep.values():
+            lpab = cmf_msgs.ReconfigurationResponse.deserialize(r)[0]
+            latest_pruneable_blocks += [lpab.response]
+
+        await op.prune(latest_pruneable_blocks)
+        # Validate that the root values are in sync after the pruning has finished
+        await bft_network.wait_for_replicas_rvt_root_values_to_be_in_sync(bft_network.all_replicas())
```
Comment (glevkovich) [inline tests/apollo/test_skvbc_state_transfer.py:357]: nit: can you please add a line break?

**Comment by glevkovich** [comment]
A general comment about commit 2d77885ff5710904e6168d91dee8540b83a2ae67, and f6dba61b39f65e7721529eed4b7d3147733de8eb : + similar ones: You are not supposed to have commits like that, as there is no real change here. It should be merged into relevant commits. User of the reporistory doesn't care about your ongoing changes while testing you code. Only significant changes should have a seperate commit.

File: tests/apollo/test_skvbc_state_transfer.py [hunk 5]
```diff
@@ -279,5 +279,74 @@ async def test_state_transfer_rvt_validity_after_pruning(self, bft_network):
...
+
+        1) Start all replicas in a given BFT network
+        2) Loop 6 times:
+            3) Send enough requests to trigger 2 checkpoints
+            4) After the 1st iteration, do pruning at each even iteration (i.e. i % 2 == 0)
+            5) Randomly choose if a replica will be restarted. If so, select a random replica and restart it
+        6) Wait for the RVT root values to be in sync
+        """
+
+        [bft_network.start_replica(i) for i in bft_network.all_replicas()]
```
Comment (glevkovich) [inline tests/apollo/test_skvbc_state_transfer.py]: nit: why the square brackets (list created for?..)

File: tests/apollo/test_skvbc_state_transfer.py [hunk 6]
```diff
@@ -279,5 +279,74 @@ async def test_state_transfer_rvt_validity_after_pruning(self, bft_network):
...
+                            break
+                print('Done pruning.')
+        
+            restart = random.choice([0, 1])
+            if restart == 1:
+                print('Selecting a random replica to be restarted (the primary is excluded)...')
+                replica_to_restart = random.choice(bft_network.all_replicas(without={0}))
+                print(f'Replica {replica_to_restart} will be restarted.')
+                bft_network.stop_replica(replica_to_restart)
+                await trio.sleep(seconds=2)
```
Comment (glevkovich) [inline tests/apollo/test_skvbc_state_transfer.py]: the next 3 lines gives us 24 seconds added to test.
We can save it, and I think we can be "more brutal" here - kill the replicas with force_kill=True

**Review by glevkovich — APPROVED** [review]
LGTM! Thanks @StefanPalashev , these are high quality tests - will help us make sure ST is not broken on master.
I"ve added few comments in code and on the PR conversation (main page), please let me know if you want to discuss any. Two of the commits can be merged fixedup) with interactive rebase, they are not really nessacary.

File: tests/apollo/util/bft.py (hunk 1 above)
Comment (StefanPalashev) [reply tests/apollo/util/bft.py]: Done, thanks.

File: tests/apollo/test_skvbc_state_transfer.py (hunk 6 above)
Comment (StefanPalashev) [reply tests/apollo/test_skvbc_state_transfer.py]: The suggested change has been applied, however a sleep of 1 second is left for the replica to start.

File: tests/apollo/test_skvbc_state_transfer.py (hunk 5 above)
Comment (StefanPalashev) [reply tests/apollo/test_skvbc_state_transfer.py]: There was no need to use list comprehension, thanks. The replicas will now be started in a regular for loop.

File: tests/apollo/util/bft.py (hunk 3 above)
Comment (StefanPalashev) [reply tests/apollo/util/bft.py]: There is no difference, the intent was to log as soon as the function is called.

**Pull request merged**